File: android/src/main/java/com/margelo/nitro/audioanvil/SpeakerWindowAssembler.kt
package com.margelo.nitro.audioanvil

import com.margelo.nitro.core.ArrayBuffer
import kotlin.math.sqrt

/** Assembles overlapping fixed-length windows from the PCM stream for speaker embedding. */
internal class SpeakerWindowAssembler(
  private val sampleRate: Int,
  windowMs: Double,
  hopMs: Double,
  private val onWindow: (SpeakerWindow) -> Unit,
) {
  private val windowSamples = maxOf(1, (sampleRate * windowMs / 1000.0).toInt())
  private val hopSamples = maxOf(1, (sampleRate * hopMs / 1000.0).toInt())
  private val ring = ShortArray(windowSamples)
  private var filled = 0
  private var writeIndex = 0
  private var samplesSinceEmit = 0
  private var endSamples = 0L
  private var primed = false

  /** `mediaSamples` is the timeline position of the first sample in `samples`. */
  fun append(samples: ShortArray, count: Int, mediaSamples: Long) {
    if (filled == 0) {
      endSamples = mediaSamples
    }
    for (i in 0 until count) {
      ring[writeIndex] = samples[i]
      writeIndex = (writeIndex + 1) % windowSamples
      if (filled < windowSamples) filled++
    }
    endSamples += count
    samplesSinceEmit += count
    if (filled < windowSamples) return
    if (!primed || samplesSinceEmit >= hopSamples) {
      primed = true
      samplesSinceEmit = 0
      emit()
    }
  }

  /** Drops buffered audio so a window never spans a pause or interruption. */
  fun reset() {
    filled = 0
    writeIndex = 0
    samplesSinceEmit = 0
    primed = false
  }

  private fun emit() {
    val bytes = ByteArray(windowSamples * 2)
    var sumSquares = 0.0
    for (i in 0 until windowSamples) {
      val sample = ring[(writeIndex + i) % windowSamples]
      val value = sample.toDouble()
      sumSquares += value * value
      bytes[i * 2] = (sample.toInt() and 0xFF).toByte()
      bytes[i * 2 + 1] = ((sample.toInt() shr 8) and 0xFF).toByte()
    }
    val rms = sqrt(sumSquares / windowSamples) / 32768.0
    val buffer = ArrayBuffer.allocate(bytes.size)
    buffer.getBuffer(false).put(bytes)
    val startSamples = endSamples - windowSamples
    onWindow(
      SpeakerWindow(
        buffer = buffer,
        startMs = startSamples / sampleRate.toDouble() * 1000.0,
        endMs = endSamples / sampleRate.toDouble() * 1000.0,
        rms = rms,
      )
    )
  }
}

File: ios/SpeakerWindowAssembler.swift
import Foundation
import NitroModules

/// Assembles overlapping fixed-length windows from the PCM stream for speaker embedding.
final class SpeakerWindowAssembler {
  private let sampleRate: Int
  private let windowSamples: Int
  private let hopSamples: Int
  private var ring: [Int16] = []
  private var samplesSinceEmit = 0
  private var endSamples = 0
  private var primed = false
  private let onWindow: (SpeakerWindow) -> Void

  init(sampleRate: Int, windowMs: Double, hopMs: Double, onWindow: @escaping (SpeakerWindow) -> Void) {
    self.sampleRate = sampleRate
    self.windowSamples = max(1, Int(Double(sampleRate) * windowMs / 1000.0))
    self.hopSamples = max(1, Int(Double(sampleRate) * hopMs / 1000.0))
    self.onWindow = onWindow
    ring.reserveCapacity(windowSamples)
  }

  /// `mediaSamples` is the timeline position of the first sample in `pcm`.
  func append(_ pcm: Data, mediaSamples: Int) {
    let samples = pcm.anvilInt16Samples
    if ring.isEmpty {
      endSamples = mediaSamples
    }
    ring.append(contentsOf: samples)
    endSamples += samples.count
    if ring.count > windowSamples {
      ring.removeFirst(ring.count - windowSamples)
    }
    samplesSinceEmit += samples.count
    guard ring.count == windowSamples else { return }
    if !primed || samplesSinceEmit >= hopSamples {
      primed = true
      samplesSinceEmit = 0
      emit()
    }
  }

  /// Drops buffered audio so a window never spans a pause or interruption.
  func reset() {
    ring.removeAll(keepingCapacity: true)
    samplesSinceEmit = 0
    primed = false
  }

  private func emit() {
  var sumSquares = 0.0
  for sample in ring { let value = Double(sample); sumSquares += value * value }
  let rms = (sumSquares / Double(ring.count)).squareRoot() / 32768.0
  let data = ring.withUnsafeBufferPointer { Data(buffer: $0) }
  let startSamples = endSamples - windowSamples
  do {
    let buffer = try ArrayBuffer.copy(data: data)
    onWindow(SpeakerWindow(
      buffer: buffer,
      startMs: Double(startSamples) / Double(sampleRate) * 1000.0,
      endMs: Double(endSamples) / Double(sampleRate) * 1000.0,
      rms: rms
    ))
  } catch {
    // Drop this window — next hop will emit again.
  }
}
}
